store/cachekv/internal/memiterator.go

```diff
@@ -0,0 +1,112 @@
+package internal
+
+import (
+	"bytes"
+
+	"github.com/cosmos/cosmos-sdk/store/types"
+	"github.com/tidwall/btree"
+)
+
+var _ types.Iterator = &memIterator{}
+
+// memIterator iterates over iterKVCache items.
+// if key is nil, means it was deleted.
+// Implements Iterator.
+type memIterator struct {
+	iter btree.GenericIter[item]
+
+	start     []byte
+	end       []byte
+	ascending bool
+	lastKey   []byte
+	deleted   map[string]struct{}
+	valid     bool
+}
+
+func NewMemIterator(start, end []byte, items *BTree, deleted map[string]struct{}, ascending bool) *memIterator {
+	iter := items.tree.Iter()
+	var valid bool
+	if ascending {
+		if start != nil {
+			valid = iter.Seek(newKey(start))
+		} else {
+			valid = iter.First()
+		}
+	} else {
+		if end != nil {
+			valid = iter.Seek(newKey(end))
+			if !valid {
+				valid = iter.Last()
+			} else {
+				// end is exclusive
+				valid = iter.Prev()
+			}
+		} else {
+			valid = iter.Last()
+		}
+	}
+	return &memIterator{
+		iter:      iter,
+		start:     start,
+		end:       end,
+		ascending: ascending,
+		lastKey:   nil,
+		deleted:   deleted,
+		valid:     valid,
+	}
+}
+
+func (mi *memIterator) Domain() (start []byte, end []byte) {
+	return mi.start, mi.end
+}
+
+func (mi *memIterator) Close() error {
+	mi.iter.Release()
+	return nil
+}
+
+func (mi *memIterator) Error() error {
+	return nil
+}
+
+func (mi *memIterator) Valid() bool {
+	if !mi.valid {
+		return false
+	}
+	key := mi.iter.Item().key
```

Comment:
Let's use the methods put out here and do
```go
key := mi.Key()
```

Reply:
done